src/transport/CASESession.cpp

```diff
@@ -121,81 +276,265 @@ CHIP_ERROR CASESession::DeriveSecureSession(SecureSession & session)
+        for (size_t i = 0; i < n_trusted_roots; ++i)
+        {
+            bbuf.Put(TrustedRoots[i], sizeof(TrustedRootIdentifier));
+        }
+        bbuf.Put(mEphemeralKey.Pubkey(), mEphemeralKey.Pubkey().Length());
+        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
+    }
 
     msg_R1->SetDataLength(data_len);
+
+    err = mCommissioningHash.AddData(msg_R1->Start(), msg_R1->DataLength());
+    SuccessOrExit(err);
+
     mNextExpectedMsg = Protocols::SecureChannel::MsgType::CASE_SigmaR2;
 
     // Call delegate to send the msg to peer
-    ReturnErrorOnFailure(AttachHeaderAndSend(Protocols::SecureChannel::MsgType::CASE_SigmaR1, std::move(msg_R1)));
+    err = AttachHeaderAndSend(Protocols::SecureChannel::MsgType::CASE_SigmaR1, std::move(msg_R1));
+    SuccessOrExit(err);
 
     ChipLogDetail(Inet, "Sent SigmaR1 msg");
 
-    return CHIP_NO_ERROR;
+exit:
+    return err;
 }
 
 CHIP_ERROR CASESession::HandleSigmaR1_and_SendSigmaR2(const PacketHeader & header, const System::PacketBufferHandle & msg)
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
 
-    uint16_t data_len = 0;
+    err = HandleSigmaR1(header, msg);
+    SuccessOrExit(err);
 
-    const uint8_t * buf = msg->Start();
+    err = SendSigmaR2();
+    SuccessOrExit(err);
 
-    System::PacketBufferHandle msg_R2;
+exit:
+    return err;
+}
 
-    ChipLogDetail(Inet, "Received SigmaR1 msg");
+CHIP_ERROR CASESession::HandleSigmaR1(const PacketHeader & header, const System::PacketBufferHandle & msg)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    const uint8_t * buf   = msg->Start();
+    uint16_t buflen       = msg->DataLength();
+    uint16_t fixed_buflen = kSigmaParamRandomNumberSize + sizeof(uint16_t) + sizeof(TrustedRoots) + kP256_PublicKey_Length;
 
     VerifyOrExit(buf != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    VerifyOrExit(buflen >= fixed_buflen, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    ChipLogDetail(Inet, "Received SigmaR1 msg");
+
+    err = mCommissioningHash.AddData(msg->Start(), msg->DataLength());
+    SuccessOrExit(err);
 
-    // TODO: Verify received SigmaR1 message.
-    // https://github.com/project-chip/connectedhomeip/issues/4469
+    {
+        // Let's skip the random number portion of the message
+        const uint8_t * msgptr = &buf[kSigmaParamRandomNumberSize];
+        // Omitting SessionID here - Mirroring PASESession, which omits SessionID as well
+        uint32_t n_trusted_roots = chip::Encoding::LittleEndian::Read16(msgptr);
+        // Step 1/2
+        // obtain certificate(s)
+        // this is dummy...
+        msgptr += kRootOfTrustCertSize * n_trusted_roots;
+        // write public key from message
+        Encoding::LittleEndian::BufferWriter bbuf(&*mRemotePubKey, mRemotePubKey.Length());
+        bbuf.Put(msgptr, kP256_PublicKey_Length);
+        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
+    }
+
+    if (header.GetSourceNodeId().HasValue() && mConnectionState.GetPeerNodeId() == kUndefinedNodeId)
```

Comment:
This logic of assigning the node ID was a temporary work around in PASE Session state machine. The Node ID will be assigned as part of device attestation. Assuming DA code gets merged before CASE Session, we may not need this code here.